File: SproInstall/SocketPro/samples/RemoteSum/csharp/Client/SumClient.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using SocketProAdapter.ClientSide;

namespace Client
{
    public partial class SumClient : Form, IAsyncResultsHandler
    {
        public SumClient()
        {
            InitializeComponent();
        }

        private CClientSocket m_cs = new CClientSocket();
        private CAsyncServiceHandler m_ash;
        private void SumClient_Load(object sender, EventArgs e)
        {
            m_ash = new CAsyncServiceHandler(RemoteSumConst.sidRemSum, m_cs, this);
            
            m_cs.m_OnSocketConnected = delegate(int hSocket, int nErrorCode)
            {
                if (nErrorCode == 0)
                {
                    btnDoSum.Enabled = true;
                    btnPause.Enabled = true;
                    btnRedoSum.Enabled = true;
                    m_cs.SwitchTo(m_ash);
                }
                else
                    MessageBox.Show(m_cs.GetErrorMsg());
            };

            m_cs.m_OnSocketClosed = delegate(int hSocket, int nErrorCode)
            {
                btnDoSum.Enabled = false;
                btnPause.Enabled = false;
                btnRedoSum.Enabled = false;
            };

            m_cs.Connect("localhost", 20901);
        }

        #region IAsyncResultsHandler Members

        public void OnExceptionFromServer(CAsyncServiceHandler AsyncServiceHandler, SocketProAdapter.CSocketProServerException Exception)
        {
            throw new Exception("The method or operation is not implemented.");
        }

        public void Process(CAsyncResult AsyncResult)
        {
            switch (AsyncResult.RequestId)
            {
                case RemoteSumConst.idPauseRemSum:
                case RemoteSumConst.idRedoSumRemSum:
                case RemoteSumConst.idDoSumRemSum:
                    {
                        int rtn = 0;
                        AsyncResult.UQueue.Pop(out rtn);
                        txtSum.Text = rtn.ToString();
                    }
                    break;
                case RemoteSumConst.idReportProgress:
                    {
                        int nWhere = 0;
                        int nSum = 0;
                        AsyncResult.UQueue.Pop(out nWhere);
                        AsyncResult.UQueue.Pop(out nSum);
                        txtSum.Text = "Where = " + nWhere.ToString() + ", Sum = " + nSum.ToString();
                    }
                    break;
                default:
                    break;
            }
        }

        #endregion

        private void btnPause_Click(object sender, EventArgs e)
        {
            m_cs.BeginBatching();

            //send a cancel request to a remote server. Here is a big secret from SocketPro!
            m_cs.Cancel(); 

            m_ash.SendRequest(RemoteSumConst.idPauseRemSum);
            m_cs.Commit(true); //make two requests in one shot
        }

        private void btnDoSum_Click(object sender, EventArgs e)
        {
            int start = 100;
            int end = 400;
            m_ash.SendRequest(RemoteSumConst.idDoSumRemSum, start, end);
        }

        private void btnRedoSum_Click(object sender, EventArgs e)
        {
            m_ash.SendRequest(RemoteSumConst.idRedoSumRemSum);
        }
    }
}

File: SproInstall/SocketPro/samples/RemoteSum/vbnet/Client/SumClient.vb
Imports System.ComponentModel
Imports System.Text
Imports SocketProAdapter.ClientSide

Namespace Client
	Partial Public Class SumClient
		Inherits Form
		Implements IAsyncResultsHandler
		Public Sub New()
			InitializeComponent()
		End Sub

		Private m_cs As New CClientSocket()
        Private m_ash As CAsyncServiceHandler

        Private Sub OnConnected(ByVal hSocket As Integer, ByVal nErrorCode As Integer)
            If nErrorCode = 0 Then
                btnDoSum.Enabled = True
                btnPause.Enabled = True
                btnRedoSum.Enabled = True
                m_cs.SwitchTo(m_ash)
            Else
                MessageBox.Show(m_cs.GetErrorMsg())
            End If
        End Sub

        Private Sub OnSocketClosed(ByVal hSocket As Integer, ByVal nErrorCode As Integer)
            btnDoSum.Enabled = False
            btnPause.Enabled = False
            btnRedoSum.Enabled = False
        End Sub

        Private Sub SumClient_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
            m_ash = New CAsyncServiceHandler(RemoteSumConst.sidRemSum, m_cs, Me)
            m_cs.m_OnSocketConnected = AddressOf OnConnected
            m_cs.m_OnSocketClosed = AddressOf OnSocketClosed
            m_cs.Connect("localhost", 20901)
        End Sub

#Region "IAsyncResultsHandler Members"

        Public Sub OnExceptionFromServer(ByVal AsyncServiceHandler As CAsyncServiceHandler, ByVal Exception As SocketProAdapter.CSocketProServerException) Implements IAsyncResultsHandler.OnExceptionFromServer
            Throw New Exception("The method or operation is not implemented.")
        End Sub

        Public Sub Process(ByVal AsyncResult As CAsyncResult) Implements IAsyncResultsHandler.Process
            Select Case AsyncResult.RequestId
                Case RemoteSumConst.idPauseRemSum, RemoteSumConst.idRedoSumRemSum, RemoteSumConst.idDoSumRemSum
                    Dim rtn As Integer = 0
                    AsyncResult.UQueue.Pop(rtn)
                    txtSum.Text = rtn.ToString()
                Case RemoteSumConst.idReportProgress
                    Dim nWhere As Integer = 0
                    Dim nSum As Integer = 0
                    AsyncResult.UQueue.Pop(nWhere)
                    AsyncResult.UQueue.Pop(nSum)
                    txtSum.Text = "Where = " & nWhere.ToString() & ", Sum = " & nSum.ToString()
                Case Else
            End Select
        End Sub

#End Region

        Private Sub btnPause_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnPause.Click
            m_cs.BeginBatching()

            'send a cancel request to a remote server. Here is a big secret from SocketPro!
            m_cs.Cancel()

            m_ash.SendRequest(RemoteSumConst.idPauseRemSum)
            m_cs.Commit(True) 'make two requests in one shot
        End Sub

        Private Sub btnDoSum_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnDoSum.Click
            Dim start As Integer = 100
            Dim last As Integer = 400
            m_ash.SendRequest(RemoteSumConst.idDoSumRemSum, start, last)
        End Sub

        Private Sub btnRedoSum_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnRedoSum.Click
            m_ash.SendRequest(RemoteSumConst.idRedoSumRemSum)
        End Sub
    End Class
End Namespace